File: src/4/4.py
import os
import psutil
import time

def print_memory_usage(note):
    process = psutil.Process(os.getpid())
    mem = process.memory_info().rss / (1024 * 1024)  # Convert bytes to MB
    print(f"[{note}] Memory usage: {mem:.2f} MB")

print_memory_usage("Start")

# Allocate a large list
arr = [i for i in range(1000000)]
print_memory_usage("After allocation")

# Pause to observe memory before deletion
time.sleep(1)

# Delete the list and force garbage collection
del arr
time.sleep(1)
print_memory_usage("After deletion")

# Ouput
# Start] Memory usage: 103.16 MB
# [After allocation] Memory usage: 141.42 MB
# [After deletion] Memory usage: 105.19 MB

File: src/4/4.java
public class MemoryDemo {
    public static void main(String[] args) throws InterruptedException {
        printMemoryUsage("Start");
        
        // Allocate an array of 1,000,000 integers
        int[] arr = new int[1000000];
        printMemoryUsage("After allocation");
        
        // Remove reference and suggest garbage collection
        arr = null;
        System.gc();
        Thread.sleep(1000);  // Wait for GC to complete
        printMemoryUsage("After deletion");
    }
    
    public static void printMemoryUsage(String note) {
        Runtime runtime = Runtime.getRuntime();
        long usedMemory = (runtime.totalMemory() - runtime.freeMemory()) / (1024 * 1024);
        System.out.println("[" + note + "] Memory usage: " + usedMemory + " MB");
    }
}

// Output
// [Start] Memory usage: 2 MB
// [After allocation] Memory usage: 6 MB
// [After deletion] Memory usage: 2 MB
